mavlink/message.go

```diff
@@ -102,53 +103,91 @@ func newPacketFromBytes(b []byte) (*Packet, int) {
 // a message they're interested in, and convert it to the
 // corresponding type via Message.FromPacket()
 func (dec *Decoder) Decode() (*Packet, error) {
-
-	// discard bytes until our start byte
 	for {
-		c, err := dec.br.ReadByte()
-		if err != nil {
-			return nil, err
+		startFoundInBuffer := false
+		// discard bytes in buffer before start byte
+		for i, b := range dec.buffer {
+			if b == startByte {
+				dec.buffer = dec.buffer[i:]
+				startFoundInBuffer = true
+				break
+			}
 		}
-		if c == startByte {
-			break
+
+		// if start not found, read until we see start byte
+		if !startFoundInBuffer {
+			for {
+				c, err := dec.br.ReadByte()
+				if err != nil {
+					return nil, err
+				}
+				if c == startByte {
+					dec.buffer = append(dec.buffer, c)
+					break
+				}
+			}
 		}
-	}
 
-	// hdr contains LENGTH, SEQ, SYSID, COMPID, MSGID
-	hdr := make([]byte, 5)
-	if _, err := io.ReadFull(dec.br, hdr); err != nil {
-		return nil, err
-	}
+		if len(dec.buffer) < 2 {
+			// read length byte
+			bytesRead := make([]byte, 1)
+			n, err := io.ReadAtLeast(dec.br, bytesRead, 1)
+			if err != nil {
+				return nil, err
+			}
 
-	p, payloadLen := newPacketFromBytes(hdr)
+			dec.buffer = append(dec.buffer, bytesRead[:n]...)
+		}
 
-	crc := x25.New()
-	crc.Write(hdr)
+		// buffer[1] is LENGTH and we've already read len(buffer) bytes
+		payloadLen := int(dec.buffer[1])
+		packetLen := hdrLen + payloadLen + numChecksumBytes
+		bytesNeeded := packetLen - len(dec.buffer)
+		if bytesNeeded > 0 {
+			bytesRead := make([]byte, bytesNeeded)
+			n, err := io.ReadAtLeast(dec.br, bytesRead, bytesNeeded)
+			if err != nil {
+				return nil, err
+			}
+			dec.buffer = append(dec.buffer, bytesRead[:n]...)
```

Comment:
May make sense to extract this pattern into a `appendAtLeast` function; occurs twice here.